getClientRects returning wrong values on empty selection

vitorroriz opened the pull request:
#### b9508a969a49f752996bd8d3dd86fd29339e004d
<pre>
getClientRects returning wrong values on empty selection
<a href="https://bugs.webkit.org/show_bug.cgi?id=148532">https://bugs.webkit.org/show_bug.cgi?id=148532</a>
&lt;rdar://22634654&gt;

Reviewed by Antti Koivisto.

* Source/WebCore/layout/integration/inline/InlineIteratorTextBox.cpp:
(WebCore::InlineIterator::TextBox::selectionRect const):

getClientRects should return only the rect at the caret position when selection is empty (or unitary range).
In this PR we changed the way to evaluate if the caret is within the text box regarding its upper limit.
If the caret position is equal the end of the text box, the caret should be considered in the range of the
text box only for the last text box in the chain. This is because we allow the caret to move to a position after
the last text box in the chain.

For the non-caret case we keep calculating if range is within the text box as before.

* Source/WebCore/layout/integration/inline/InlineIteratorTextBox.cpp:
(WebCore::InlineIterator::TextBox::selectionRect const):
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::localQuadForTextRun):

We are now passing the real offset limits (start, end) of the range to TextBox::selectionRect so
it can identify a caret case. This seems to be unnecessary since we are clamping the range limits in relation
to the textbox inside selectionRect. Also, without that we can&apos;t differ between a caret case and a case in which the text box
has just 1 position overlapping with.

Canonical link: <a href="https://commits.webkit.org/253536@main">https://commits.webkit.org/253536@main</a>
</pre>

nt1m commented on Source/WebCore/layout/integration/inline/InlineIteratorTextBox.cpp:
```diff
@@ -41,10 +41,25 @@ TextBoxIterator TextBox::nextTextBox() const
+        auto isLastTextBox = nextTextBox() == itEnd;
+
+        if ((isLastTextBox && !isCaretWithinLastTextBox())
+            || (!isLastTextBox && !isCaretWithinTextBox()))
```
this fits in one line

anttijk commented on Source/WebCore/layout/integration/inline/InlineIteratorTextBox.cpp:
```diff
@@ -41,10 +41,25 @@ TextBoxIterator TextBox::nextTextBox() const
+    // For last text box in a InlineTextBox chain, we allow the caret to move to a position 'after' the end of the last text box.
+    auto isCaretWithinLastTextBox = [&]() {
+        return rangeStart >= start() && rangeStart <= end();
+    };
```
These could be just bools, lambdas don't really add anything here.

anttijk reviewed (approved):
You should explain the fix in the commit message.

anttijk commented on Source/WebCore/layout/integration/inline/InlineIteratorTextBox.cpp:
```diff
@@ -41,10 +41,29 @@ TextBoxIterator TextBox::nextTextBox() const
-    if (clampedStart >= clampedEnd && !(rangeStart == rangeEnd && rangeStart >= start() && rangeStart <= end()))
-        return { };
+    if (clampedStart >= clampedEnd) {
+        if (!isCaretCase) {
```
There is no obvious reason to use negated test here. Maybe just flip the branches around?

vitorroriz replied:
Sure, fixed.

anttijk commented on Source/WebCore/layout/integration/inline/InlineIteratorTextBox.cpp:
```diff
@@ -41,10 +41,29 @@ TextBoxIterator TextBox::nextTextBox() const
 LayoutRect TextBox::selectionRect(unsigned rangeStart, unsigned rangeEnd) const
 {
+    bool isCaretCase = rangeStart == rangeEnd;
+    rangeEnd = std::min(end(), rangeEnd);
```
It would be good to understand why this is needed. Can it be handled by clamp()?

vitorroriz replied:
I also think we don't since we are clamping anyway. I've removed it and adjusted the checks for caret within box. It still passes our Range tests.

vitorroriz committed:
getClientRects returning wrong values on empty selection
https://bugs.webkit.org/show_bug.cgi?id=148532
<rdar://22634654>

Reviewed by Antti Koivisto.

* Source/WebCore/layout/integration/inline/InlineIteratorTextBox.cpp:
(WebCore::InlineIterator::TextBox::selectionRect const):

getClientRects should return only the rect at the caret position when selection is empty (or unitary range).
In this PR we changed the way to evaluate if the caret is within the text box regarding its upper limit.
If the caret position is equal the end of the text box, the caret should be considered in the range of the
text box only for the last text box in the chain. This is because we allow the caret to move to a position after
the last text box in the chain.

For the non-caret case we keep calculating if range is within the text box as before.

* Source/WebCore/layout/integration/inline/InlineIteratorTextBox.cpp:
(WebCore::InlineIterator::TextBox::selectionRect const):
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::localQuadForTextRun):

We are now passing the real offset limits (start, end) of the range to TextBox::selectionRect so
it can identify a caret case. This seems to be unnecessary since we are clamping the range limits in relation
to the textbox inside selectionRect. Also, without that we can't differ between a caret case and a case in which the text box
has just 1 position overlapping with.

Canonical link: https://commits.webkit.org/253536@main
